### PR TITLE
Add additional rounding strategies

### LINKED ISSUE #140: Add support for additional rounding strategies
Hi, first of all thanks for the library,I just started using it and the API is excellent, really easy to get started with :slightly_smiling_face: 

For my use case, I need to make use of the round half up rounding strategy, I was wondering if you were open to adding the ability to either:
 - use more than banker's rounding
 - provide a function that acts as the rounding strategy

I'm more than willing to attempt implementing this, I just wanted to see if it was something you were open to adding into the crate?

### PR DESCRIPTION
**Pull request by AndrewSpeed**
This PR implements the rounding strategies discussed in #140, namely `RoundingStrategy::RoundHalfUp` and `RoundingStrategy::RoundHalfDown`.

In order to do so, it maps the existing strategy to `RoundingStrategy::BankersRounding`.

Resolves #140

### CODE REVIEW

File: src/decimal.rs [hunk 1]
```diff
@@ -514,116 +528,168 @@ impl Decimal {
...
+
+        // If the decimal_portion is zero then we round based on the other data
+        let mut cap = [5, 0, 0];
+        for _ in 0..(old_scale - dp - 1) {
+            mul_by_u32(&mut cap, 10);
+        }
+        let order = cmp_internal(&decimal_portion, &cap);
+
+
+        match strategy {
```
Comment (AndrewSpeed) [inline src/decimal.rs]: I'm thinking about trying to refactor strategies into a trait, that way this crate can provide some default implementations, but users can extend it further if they wish to do so.

If this approach is taken, I think some additional methods would need to be exposed to make it possible for alternative strategies to be implemented.

File: tests/decimal_tests.rs [hunk 2]
```diff
@@ -648,6 +648,275 @@ fn it_can_round_to_2dp() {
     assert_eq!("6.12", b.to_string());
 }
 
+#[test]
+fn it_can_round_to_2dp_using_bankers_rounding_strategy() {
+    let a = Decimal::from_str("6.12345").unwrap();
```
Comment (paupino) [inline tests/decimal_tests.rs]: On a side note, I've been more inclined to use test arrays going forward simply because it makes it easier to add tests. I think you could do probably set up something similar here to set precedent:
```
using RoundingStrategy::*;
let tests = &[
    ("6.12345", 2, BankersRounding, "6.12"),
    ("6.126", 2, BankersRounding, "6.13"),
    // etc
];
for &(input, dp, strategy, expected) in tests {
    let a = Decimal::from_str(input).unwrap();
    let b = a.round_dp_with_strategy(dp, strategy);
    assert_eq!(expected, b.to_string());
}
```
The odd one out of course is the test marked `// Issue #71`. For this one it could have a separate combined test which iterates the various rounding strategies on this irrational number.

Of course I realize that you're reusing an existing pattern - it's just one I'd like to move away from at some stage in the future. Whether that's now or later is up for debate 😄

File: src/decimal.rs [hunk 3]
```diff
@@ -514,116 +528,168 @@ impl Decimal {
...
-            // Short circuit for zero
-            if self.is_zero() {
-                return Decimal {
-                    lo: 0,
-                    mid: 0,
-                    hi: 0,
-                    flags: flags(self.is_sign_negative(), dp),
-                };
-            }
+        let old_scale = self.scale();
```
Comment (paupino) [inline src/decimal.rs]: Any reason you do this after the previous comparison? As it is it just means the flags get evaluated twice.

File: src/decimal.rs (hunk 1 above)
Comment (paupino) [reply src/decimal.rs]: Interesting idea. I like the extensibility aspect of this - how would you ensure it remained ergonomic?

Comment (AndrewSpeed) [reply src/decimal.rs]: Assuming you're talking about ergonomics for the method and/or trait, my thought was that we could have this method call out to a method defined on the trait e.g. `round_to_dp(n: u32)`, from there it's up to the strategy on _how_ to do the rounding.

This should give the solution the most flexibility, as the implementation won't know anything about the rounding, except that it returns a `Decimal`.

In summary, something like:
```rust
trait RoundingStrategy {...}

pub fn round_dp_with_strategy(&self, dp: u32, strategy: RoundingStrategy) -> Decimal {
  strategy.round_to_dp(self, dp)
}
```

Can you see any problems and/or concerns with that approach?

Comment (paupino) [reply src/decimal.rs]: Playing it out from an integration perspective; are you thinking:
```
let number = dec!(1.2345);
let _ = number.round_with_strategy(2, BankersRoundingStrategy::default());
```

You could also feasibly invert this relationship also:
```
let number = dec!(1.2345);
let bankers_rounding = BankersRounding::with_dp(2);
let _ = bankers_rounding.round(number);
```

The problem I see with each is accessing/setting the internals effectively for efficiency. I think that at the moment, an `enum` is great for optimizing towards efficiency because it allows everything to be handled within the same scope. If another rounding strategy is required later on then it is feasible for someone to do this using the public API of the decimal type in an approach similar to the second example (at a small perf cost).

On a side note, the function name is starting to get quite long and since it'll become part of the public API is the perfect time to think about it. I've been trying to think of alternatives to assist with this but the best I can come up with is _maybe_ dropping the `_dp`. I'm still a little torn here... I'm definitely open to keeping it the same but if you think of something more succinct then I'm all ears!

Comment (AndrewSpeed) [reply src/decimal.rs]: That's along the lines of what I was thinking of, however another alternative would be to set the rounding strategy when the decimal is created. `BankersRounding` would be the default, then the user could update it or use a specialised constructor like:

```rust
let number = Decimal::new_with_strategy(134, 2, Rounding::RoundHalfUp);
let rounded = number.round_dp(2);
```

This would eliminate the naming problem entirely 🙂 

I can definitely see where you're coming from, it'd be a trade-off of performance for flexibility, I'm just not sure what that impact would be, and whether it'd be worth it in the long run.

**Comment by AndrewSpeed** [comment]
Sorry for the delay in updating this PR @paupino I've added the feedback you requested. Let me know if there's anything else that you'd like changed :slightly_smiling_face:

**Comment by paupino** [comment]
Thank you @AndrewSpeed ! I see you refactored the `BankersRounding` and `RoundHalfUp` tests to use the array style which is looking great. `RoundHalfDown` tests are still in the old format though - could these be also switched over?

**Comment by AndrewSpeed** [comment]
:+1: Will have a commit up to address that shortly

**Comment by paupino** [comment]
Awesome, thank you for bearing with me through that PR! I'll merge these once the tests pass.

**Comment by AndrewSpeed** [comment]
No problem at all, sorry for the delay between feedback and it being addressed, life happened and I got bogged down in other things :sweat_smile:
